### PR TITLE
Showcase the changes present in `neuropoly-fork-patches`

### PR DESCRIPTION
**Pull request by joshuacwnewton**
> [!CAUTION]
> This PR should not be merged. The purpose of this PR is simply to showcase the changes currently on the `neuropoly-fork-patches` branch. 

## instructions

- This branch should be continuously updated against `master` (to ensure that the changes in this branch are always compatible with the most recent version of `nnunetv2`)
   ```bash
   # prerequisite: sync the nnUNet-neuropoly `master` with the upstream `master` on GitHub
   git checkout master
   git pull master
   git checkout neuropoly-fork-patches
   git rebase master
   # during rebase: fix any conflicts
   git push --force
- Whenever there is a new upstream `nnunetv2` release, we should create a new branch for a parallel release on our own `nnunetv2-neuropoly` project, then cherry-pick the changes from this `neuropoly-fork-changes` branch onto the release branch.

### CODE REVIEW

File: nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py [hunk 1]
```diff
@@ -58,3 +58,9 @@ def _build_loss(self):
             loss = DeepSupervisionWrapper(loss, weights)
         return loss
 
+
+class nnUNetTrainerDiceCELoss_noSmooth_4000epochs_fromScratch(nnUNetTrainerDiceCELoss_noSmooth):
+    def __init__(self, plans: dict, configuration: str, fold: int, dataset_json: dict,
+                 device: torch.device = torch.device('cuda')):
+        super().__init__(plans, configuration, fold, dataset_json, device)
+        self.num_epochs = 4000
```
Comment (joshuacwnewton) [inline nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py]: @plbenveniste - Regarding the addition of a new trainer class: We actually don't need to make this change on the fork?

----

Explanation: In this branch, you can see I added a new `trainer_class` argument: https://github.com/spinalcordtoolbox/nnUNet-neuropoly/pull/1/commits/511204ebed0a780cbdf0e14d011fc037f52f5796

What this allows you to do is put the trainer class in the `.zip` for your model. For example, in a file called `trainer_class.py`, you would do:

```python
class nnUNetTrainerDiceCELoss_noSmooth_4000epochs_fromScratch(nnUNetTrainerDiceCELoss_noSmooth):
    def __init__(self, plans: dict, configuration: str, fold: int, dataset_json: dict,
                 device: torch.device = torch.device('cuda')):
        super().__init__(plans, configuration, fold, dataset_json, device)
        self.num_epochs = 4000

# using a standardized function name so that SCT can import the class
def get_trainer_class():
   return nnUNetTrainerDiceCELoss_noSmooth_4000epochs_fromScratch
```

Then, in SCT, we can do something like this:

```python
   from nnunetv2.inference.predict_from_raw_data import nnUNetPredictor
   from lesion_ms.trainer_class import get_trainer_class

   predictor = nnUNetPredictor
   predictor.initialize_from_trained_model_folder(
      # Other args
      trainer_class=get_trainer_class()
   )
```

The purpose of this is to distribute the trainer class _with the model files_, that way it will always be up to date + in sync with the model.

File: pyproject.toml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 name = "nnunetv2"
```
Comment (joshuacwnewton) [inline pyproject.toml]: Regarding the `name` parameter in `pyproject.toml` (https://github.com/spinalcordtoolbox/nnUNet-neuropoly/pull/1/commits/932ec6882fcc9ed9869669b1f3d1f348496a896e)

The reason I set the name as `nnunetv2-neuropoly` is so that we can push the package to PyPI under a unique name: https://test.pypi.org/project/nnunetv2-neuropoly/

There is a way to fix the naming conflict. See: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/5003/commits/cd73de2e930fa551ff745ee99a2192ad588bedfa

**Comment by joshuacwnewton** [comment]
@plbenveniste I think these are the next steps to take:

- [x] Restore the change `name="nnunetv2-neuropoly"`
- [x] Create a new "developer" release on "[Test PyPI](https://test.pypi.org/project/nnunetv2-neuropoly/#history)" for this branch (containing your `logits` fixes)
  - Since my last "developer" release was called "[2.6.2.dev1](https://test.pypi.org/project/nnunetv2-neuropoly/2.6.2.dev1/)", the next release can be called "2.6.2.dev2" 
  - The commands I used were just:
     ```bash
     python3 -m build
     python3 -m twine upload --repository testpypi dist/*
     ```
- [x] Update https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/5027 to use this release.
  - You can copy all of the changes from https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/5003.
  - Some of these changes are temporary "hacks" because of `totalspineseg`.

Once we confirm that the Test PyPI release works with your PR, and we are happy with the changes, then we can make some more permanent changes:

- [x] Update the release of `lesion_ms` to include the "trainer class" file in the `.zip`.
- [ ] Release `nnunetv2-neuropoly` to PyPI instead of Test PyPI
- [ ] Update `totalspineseg` to use `nnunetv2-neuropoly`.
- [ ] Update the SCT PR to remove all of the "hacky" fixes.
- [ ] Update the SCT PR to import the trainer class from the model files.

File: nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py (hunk 1 above)
Comment (plbenveniste) [reply nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py]: @joshuacwnewton Do you want me to remove this change as well and add the trainer to the model.zip? Or should we keep it like this?

File: pyproject.toml (hunk 2 above)
Comment (plbenveniste) [reply pyproject.toml]: Understood!

File: nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py (hunk 1 above)
Comment (joshuacwnewton) [reply nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py]: You can leave the trainer class here for now (just for testing). It will be useful for testing the fork on your SCT branch.

But, when we do the final release for PyPI, then we can remove the class and put it in the model `.zip` instead.

Comment (plbenveniste) [reply nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py]: Trainer class file `trainer_class.py` was added to model_fold0.zip in the release https://github.com/ivadomed/ms-lesion-agnostic/releases/tag/r20250909:

Here is the code it contains:
```python
import torch
from nnunetv2.training.nnUNetTrainer.variants.loss.nnUNetTrainerDiceLoss import nnUNetTrainerDiceCELoss_noSmooth

class nnUNetTrainerDiceCELoss_noSmooth_4000epochs_fromScratch(nnUNetTrainerDiceCELoss_noSmooth):
    def __init__(self, plans: dict, configuration: str, fold: int, dataset_json: dict,
                 device: torch.device = torch.device('cuda')):
        super().__init__(plans, configuration, fold, dataset_json, device)
        self.num_epochs = 4000

# using a standardized function name so that SCT can import the class
def get_trainer_class():
   return nnUNetTrainerDiceCELoss_noSmooth_4000epochs_fromScratch
```

**Comment by plbenveniste** [comment]
I have done the following modifications:
- changed name back to nnunet-neuropoly
- removed the trainer specific to the lesion_ms model
- pushed version 2.6.2.dev3 to Test PyPI

@joshuacwnewton Do you want me to push to PyPI or would you rather do it yourself?

**Comment by joshuacwnewton** [comment]
I think maybe we should wait to push to PyPI until we've fully tested and approved the main SCT PR? Since if we make a mistake or need to revert something, the consequences are much more costly on the main PyPI (disruptions etc.).

**Comment by plbenveniste** [comment]
> I think maybe we should wait to push to PyPI until we've fully tested and approved the main SCT PR? Since if we make a mistake or need to revert something, the consequences are much more costly on the main PyPI (disruptions etc.).

Ok good to know! I am almost pushed it PyPi and changed my mind at the very last minute. 😆

**Comment by plbenveniste** [comment]
@joshuacwnewton What should be done to move forward with this? I would like to merge the dev branch in SCT for the new MS lesion model soon

**Comment by joshuacwnewton** [comment]
Thank you for the ping, and my apologies for falling behind on this task.

- [x] Now that you have added the `trainer_class` behavior to SCT, and you have gone through the steps of putting the trainer class in the model `.zip` (and our existing test for `lesion_ms` had passed), I think it should be OK to build and push our wheel of `nnunetv2-neuropoly` to the real PyPI. (I'll do this shortly.)
  - [x] I need to clean up this branch a little to make sure we only have the necessary commits. 
  - [x] Then I need to take the changes on this branch and copy them over to the necessary version tags that we wish to have for our fork. (EDIT: Done! [`2.6.2.patch`](https://github.com/spinalcordtoolbox/nnUNet-neuropoly/releases/tag/v2.6.2.patch))
  - [x] Done! Release: https://pypi.org/project/nnunetv2-neuropoly/2.6.2 
- [ ] Once this is done, we need to update `totalspineseg` to use our fork instead of `nnunetv2` and push a new release.
- [ ] Once that is done, we can remove all of the "hacks" that we have added to the SCT side of things (e.g. `install_sct`), and instead simply specify `nnunetv2-neuropoly` in SCT's `requirements.txt` file.
- [ ] Once that's all done, your PR can be properly reviewed and merged like normal.
